node/core/approval-voting/src/lib.rs

```diff
@@ -2085,6 +2164,11 @@ async fn launch_approval(
 				// TODO: issue dispute, but not for timeouts.
 				// https://github.com/paritytech/polkadot/issues/2176
 				metrics_guard.take().on_approval_invalid();
+
+				ApprovalState::verified(
```

Comment:
Should we remove candidate hashes with invalid verifications at this point?

It seems like it is beneficial to keep track of invalidity in order to ensure we needn't verify this candidate_hash again until the candidate has been dropped by the entire validator network.

Reply:
Yep. LruCache should track something like `enum { Active(Vec<BlockHash>), Approved, Failed }`.